### PR TITLE
Wipe cache for all accounts - MSAL

### PR DESCRIPTION
**Pull request by juan-arias**
## Proposed changes

Describe what this PR is trying to do.

## Type of change

- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [x] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: MSAL/src/MSALPublicClientApplication.m [hunk 1]
```diff
@@ -1434,12 +1434,27 @@ - (void)signoutWithAccount:(nonnull MSALAccount *)account
         block(NO, localError, nil);
         return;
     }
+
+    if(signoutParameters.wipeCacheForAllAccounts)
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m:1438]: Code style: add a space after if, should be:

if (signoutParameters.wipeCacheForAllAccounts)

File: MSAL/src/MSALPublicClientApplication.m [hunk 2]
```diff
@@ -1434,12 +1434,27 @@ - (void)signoutWithAccount:(nonnull MSALAccount *)account
...
+    {
+        BOOL result = YES;
+        NSError *localError;
+        
+        result = [self.tokenCache clearCacheForAllAccountsWithContext:nil error:&localError];
+        
+        if(!result)
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m:1445]: same here - add a space after if

File: MSAL/src/public/MSALSignoutParameters.h [hunk 3]
```diff
@@ -56,6 +56,17 @@ NS_ASSUME_NONNULL_BEGIN
...
 @property (nonatomic) BOOL wipeAccount;
 
+/*
+  When flag is set, following should happen:
+    - Wipe all known universal cache locations regardless of the clientId, account etc. Should include all tokens and metadata for any cloud.
+    - Wipe all known legacy ADAL cache locations regardless of the clientId, account etc.
+    - Declare a header file with a list of additional locations for partner caches to be wiped (e.g. Teams, VisualStudio etc). Wipe operation should wipe out all those additional locations. Header file should include "display identifier" of the location (e.g. Teams cache), and precise identifiers like kSecAttrAccount, kSecAttrService etc.
+    - If SSO extension is present, call SSO extension wipe operation. Wipe operation should only be allowed to the privileged applications like Intune CP on macOS or Authenticator on iOS.
+    - Failing any of the steps should return error back to the app including exact locations and apps that failed to be cleared.
+  NO by default.
```
Comment (oldalton) [inline MSAL/src/public/MSALSignoutParameters.h:66]: Let's also add a warning saying that's "a dangerous operation" in the end :)? Great comment otherwise.

**Review by oldalton — APPROVED** [review]

File: MSAL/src/MSALPublicClientApplication.m (hunk 1 above)
Comment (juan-arias) [reply MSAL/src/MSALPublicClientApplication.m]: Fixed

File: MSAL/src/MSALPublicClientApplication.m (hunk 2 above)
Comment (juan-arias) [reply MSAL/src/MSALPublicClientApplication.m]: Fixed

File: MSAL/src/public/MSALSignoutParameters.h (hunk 3 above)
Comment (juan-arias) [reply MSAL/src/public/MSALSignoutParameters.h:66]: Added additional warning.

File: MSAL/src/MSALPublicClientApplication.m [hunk 4]
```diff
@@ -1434,12 +1435,61 @@ - (void)signoutWithAccount:(nonnull MSALAccount *)account
...
+                    {
+                        locationErrors = [[NSMutableArray alloc] init];
+                    }
+                    [locationErrors addObject:[NSString stringWithFormat:@"'%@'", key]];
+                }
+            }
+            
+            if (!result && locationErrors)
+            {
+                NSError *additionalLocationError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, [NSString stringWithFormat:@"WipeCacheForAllAccounts - error when removing cache for the item(s): %@", [locationErrors componentsJoinedByString:@", "]], nil, nil, localError, nil, nil, YES);
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m:1479]: Can you also return allLocations in userInfo as an array? Might be useful for CP UX.

File: MSAL/test/unit/MSALPublicClientApplicationTests.m [hunk 5]
```diff
@@ -3468,6 +3468,70 @@ - (void)testSignoutWithAccount_whenNonNilAccount_andWipeAccountTrue_andBrokerDis
     [self waitForExpectations:@[expectation] timeout:1];
 }
 
+- (void)testSignoutWithAccount_whenNonNilAccount_andWipeAllAccountsTrue_andBrokerDisabled_shouldWipeAllAccounts
```
Comment (oldalton) [inline MSAL/test/unit/MSALPublicClientApplicationTests.m:3471]: Would be also good to add a few tests that check wiping for additional partner locations if possible

**Review by oldalton — APPROVED** [review]
Approved with a few suggestions.

File: MSAL/src/MSALPublicClientApplication.m (hunk 4 above)
Comment (juan-arias) [reply MSAL/src/MSALPublicClientApplication.m]: Updated

File: MSAL/test/unit/MSALPublicClientApplicationTests.m (hunk 5 above)
Comment (juan-arias) [reply MSAL/test/unit/MSALPublicClientApplicationTests.m:3475]: Added additional unit test.

File: MSAL/src/MSALPublicClientApplication.m [hunk 6]
```diff
@@ -1434,12 +1435,74 @@ - (void)signoutWithAccount:(nonnull MSALAccount *)account
...
+                NSDictionary *cacheLocation = additionalPartnerLocations[locationName];
+                
+                // Try to read the keychain data in order to trigger the prompt asking for login password, user HAS TO click 'Always Allow' to then be able to delete it.
+                [keychainAccessor getDataWithAttributes:cacheLocation
+                                                context:nil
+                                                  error:&localError];
+                
+                if (localError)
+                {
+                    NSError *additionalLocationError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, [NSString stringWithFormat:@"WipeCacheForAllAccounts - error when reading cache for the item: %@.", locationName], nil, nil, localError, nil, nil, YES);
```
Comment (oldalton) [inline MSAL/src/MSALPublicClientApplication.m:1470]: Should you proceed to the next one if one fails for best effort cleanup?

**Review by oldalton — APPROVED** [review]

File: MSAL/src/MSALPublicClientApplication.m (hunk 6 above)
Comment (juan-arias) [reply MSAL/src/MSALPublicClientApplication.m:1478]: yes, this was updated

**Pull request merged**